### PR TITLE
Add ORTModule test samples for deepspeed pipeline parallel module and fairscale sharded optimizer 

### PR DESCRIPTION
**Pull request by harshithapv**
Adding ORTModule test samples for deepspeed pipeline parallel module and fairscale sharded optimizer and their corresponding CI tests.

### CODE REVIEW

File: orttraining/orttraining/test/python/orttraining_test_ortmodule_deepspeed_pipeline_parallel.py [hunk 1]
```diff
@@ -0,0 +1,121 @@
...
+torch.cuda.set_device(args.local_rank)
+device = torch.device("cuda", args.local_rank)
+
+dist.init_process_group(backend=args.backend)
+torch.manual_seed(args.seed)
+# Model.
+
+if args.run_without_ort:
+    model = nn.Sequential(
+			nn.Linear(d_in, d_hidden),     # Stage 1
```
Comment (xzhu1900) [inline orttraining/orttraining/test/python/orttraining_test_ortmodule_deepspeed_pipeline_parallel.py:74]: >			 [](start = 0, length = 3)

nit: please use 4 spaces instead of tab in code. (Same as the other instances in this file)

File: orttraining/orttraining/test/python/orttraining_test_ortmodule_deepspeed_pipeline_parallel.py [hunk 2]
```diff
@@ -0,0 +1,121 @@
...
+			nn.ReLU(),                     # Stage 2
+			nn.Linear(d_hidden, d_out)     # Stage 2
+			)
+
+
+else:
+    print("*************************************** testing with ORTModule ************")
+    model = nn.Sequential(
+			ORTModule(nn.Linear(d_in, d_hidden).to(device)),     # Stage 1
+			nn.ReLU().to(device),#ORTModule(nn.ReLU().to(device)),                     # Stage 1
```
Comment (xzhu1900) [inline orttraining/orttraining/test/python/orttraining_test_ortmodule_deepspeed_pipeline_parallel.py:88]: >,#ORTModule(nn.ReLU().to(device)), [](start = 23, length = 34)

Is there some reason we can not use ORTModule for nn.ReLu in this layer?

File: orttraining/orttraining/test/python/orttraining_test_ortmodule_deepspeed_pipeline_parallel.py [hunk 3]
```diff
@@ -0,0 +1,121 @@
...
+
+import onnxruntime
+from onnxruntime.training import ORTModule
+
+import argparse
+
+# USAGE:
+# pip install deepspeed
+# deepspeed orttraining_test_ortmodule_deepspeed_pipeline.py --deepspeed_config=orttraining_test_ortmodule_deepspeed_pipeline_config.json --pipeline-parallel-size 2 --steps=100
+# expected output : steps: 100 loss: 0.0585 iter time (s): 0.186 samples/sec: 53.694
```
Comment (xzhu1900) [inline orttraining/orttraining/test/python/orttraining_test_ortmodule_deepspeed_pipeline_parallel.py:16]: >expected output : steps: 100 loss: 0.0585 iter time (s): 0.186 samples/sec: 53.694  [](start = 2, length = 83)

Do we verify this expected output somewhere in the test?

File: orttraining/orttraining/test/python/orttraining_test_ortmodule_fairscale_sharded_optimizer.py [hunk 4]
```diff
@@ -0,0 +1,288 @@
...
+    # Data loading code
+    train_dataset = torchvision.datasets.MNIST(
+        root=TEMPDIR,
+        train=True,
+        transform=transforms.ToTensor(),
+        download=True
+    ) 
+    
+    train_sampler = torch.utils.data.distributed.DistributedSampler(
+    	train_dataset,
```
Comment (xzhu1900) [inline orttraining/orttraining/test/python/orttraining_test_ortmodule_fairscale_sharded_optimizer.py:60]: >	 [](start = 4, length = 1)

nit: please replace tabs with 4 spaces in this file.

File: orttraining/orttraining/test/python/orttraining_test_ortmodule_fairscale_sharded_optimizer.py [hunk 5]
```diff
@@ -0,0 +1,288 @@
...
+    parser.add_argument('--test-batch-size', type=int, default=64, metavar='N',
+                        help='input batch size for testing (default: 64)')
+    parser.add_argument('--no-cuda', action='store_true', default=False,
+                        help='disables CUDA training')
+    parser.add_argument('--seed', type=int, default=42, metavar='S',
+                        help='random seed (default: 42)')
+    
+    args = parser.parse_args()
+
+    #BACKEND = "nccl" if (not args.gloo or not torch.cuda.is_available()) and not args.cpu else "gloo"
```
Comment (xzhu1900) [inline orttraining/orttraining/test/python/orttraining_test_ortmodule_fairscale_sharded_optimizer.py:276]: >#BACKEND = "nccl" if (not args.gloo or not torch.cuda.is_available()) and not args.cpu else "gloo" [](start = 4, length = 98)

stale line?

File: orttraining/orttraining/test/python/orttraining_test_ortmodule_deepspeed_pipeline_parallel.py (hunk 2 above)
Comment (harshithapv) [reply orttraining/orttraining/test/python/orttraining_test_ortmodule_deepspeed_pipeline_parallel.py:88]: To instantiate an Inference session we need to know the device in which we will run it ahead of time. Since ReLU is not trainable, we don't know which device to use. So, for now, ORTModule will not handle it. Thiago is working on possible fixes in these PRs:
[1] https://github.com/microsoft/onnxruntime/pull/6365 and https://github.com/microsoft/onnxruntime/pull/6417

File: orttraining/orttraining/python/training/_utils.py [hunk 6]
```diff
@@ -43,6 +42,53 @@ def get_device_index_from_input(input):
...
+    except StopIteration:
+        # Model doesn't have a device set to any of the model parameters
+        pass
+    return device
+
+def get_device_str(device):
+    if isinstance(device, str):
+        # could be 'cuda:0', 'cuda:1', or 'cpu'. with cpu, set index=0
+        if device.find(':') == -1:
+            device += ':' + str(torch.cuda.current_device())
```
Comment (satyajandhyala) [inline orttraining/orttraining/python/training/_utils.py:72]: Is the else case an error?

**Comment by harshithapv** [comment]
Closing this PR as there are too many changes. Made a new PR with master https://github.com/microsoft/onnxruntime/pull/7078. All comments in this PR will be addressed in the new PR.
